pkg/compactor/blocks_cleaner.go

```diff
@@ -136,46 +139,79 @@ func NewBlocksCleaner(cfg BlocksCleanerConfig, bucketClient objstore.Bucket, own
 		}, []string{"user"}),
 	}
 
-	c.Service = services.NewTimerService(cfg.CleanupInterval, c.starting, c.ticker, nil)
+	c.Service = services.NewTimerService(cfg.CleanupInterval, c.starting, c.ticker, c.stopping)
 
 	return c
 }
 
+func (c *BlocksCleaner) stopping(error) error {
+	c.singleFlight.Wait()
+	return nil
+}
+
 func (c *BlocksCleaner) starting(ctx context.Context) error {
 	// Run an initial cleanup in starting state. (Note that compactor no longer waits
 	// for blocks cleaner to finish starting before it starts compactions.)
-	c.runCleanup(ctx)
+	c.runCleanup(ctx, false)
 
 	return nil
 }
 
 func (c *BlocksCleaner) ticker(ctx context.Context) error {
-	c.runCleanup(ctx)
+	c.runCleanup(ctx, true)
 
 	return nil
 }
 
-func (c *BlocksCleaner) runCleanup(ctx context.Context) {
-	level.Info(c.logger).Log("msg", "started blocks cleanup and maintenance")
+func (c *BlocksCleaner) runCleanup(ctx context.Context, async bool) {
+	// Wrap logger with some unique ID so if runCleanUp does run in parallel with itself, we can
+	// at least differentiate the logs in this function for each run.
+	logger := log.With(c.logger, "run_id", strconv.FormatInt(time.Now().Unix(), 10))
+
+	c.instrumentStartedCleanupRun(logger)
+
+	allUsers, isDeleted, err := c.refreshOwnedUsers(ctx)
+	if err != nil {
+		c.instrumentFinishedCleanupRun(err, logger)
+		return
+	}
+
+	doCleanup := func() {
+		err := c.cleanUsers(ctx, allUsers, isDeleted)
+		c.instrumentFinishedCleanupRun(err, logger)
+	}
+
+	if async {
+		go doCleanup()
+	} else {
+		doCleanup()
+	}
+}
+
+func (c *BlocksCleaner) instrumentStartedCleanupRun(logger log.Logger) {
+	logger.Log("msg", "started blocks cleanup and maintenance")
 	c.runsStarted.Inc()
+}
 
-	if err := c.cleanUsers(ctx); err == nil {
-		level.Info(c.logger).Log("msg", "successfully completed blocks cleanup and maintenance")
+func (c *BlocksCleaner) instrumentFinishedCleanupRun(err error, logger log.Logger) {
+	if err == nil {
+		logger.Log("msg", "successfully completed blocks cleanup and maintenance")
 		c.runsCompleted.Inc()
 		c.runsLastSuccess.SetToCurrentTime()
 	} else if errors.Is(err, context.Canceled) {
-		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance", "err", err)
+		logger.Log("msg", "canceled blocks cleanup and maintenance", "err", err)
 		return
 	} else {
-		level.Error(c.logger).Log("msg", "failed to run blocks cleanup and maintenance", "err", err.Error())
+		logger.Log("msg", "failed to run blocks cleanup and maintenance", "err", err.Error())
 		c.runsFailed.Inc()
 	}
 }
 
-func (c *BlocksCleaner) cleanUsers(ctx context.Context) error {
+// refreshOwnedUsers can run concurrently with itself and with a cleanup job for any tenant
```

Comment:
This comment is not clear to me. I would suggest:

```suggestion
// refreshOwnedUsers is not required to be concurrency safe, but a single instance of this function
// could run concurrently with the cleanup job for any tenant.tenant
```